python/lsst/afw/detection/_psf.cc

```diff
@@ -51,10 +52,10 @@ void wrapPsf(utils::python::WrapperCollection& wrappers) {
     wrappers.addSignatureDependency("lsst.afw.fits");
 
     auto clsPsf = wrappers.wrapType(
-            py::class_<Psf, std::shared_ptr<Psf>, typehandling::Storable>(wrappers.module, "Psf"),
+            py::class_<Psf, std::shared_ptr<Psf>, typehandling::Storable, PyPsf<>>(wrappers.module, "Psf"),
             [](auto& mod, auto& cls) {
                 table::io::python::addPersistableMethods<Psf>(cls);
-
+                cls.def(py::init<>());  // Dummy constructor for pure-Python subclasses
```

Comment:
Interesting.  My naive attempt at the factory failed since returning a `Psf` instance isn't allowed (b/c it's abstract).

Also interesting, however, is that the protected `Psf::Psf` constructor *does* get called by `Psf.__init__(self)`, as confirmed by placing a cout statement therein.  I can't see a way to call it with non-default arguments though.

So this would be ~okay so long as python subclasses can live with the `isFixed` and `capacity` defaults.

I'm looking into [Forced trampoline class initialisation](https://pybind11.readthedocs.io/en/stable/advanced/classes.html#forced-trampoline-class-initialisation) now though.  If we were ever constructing `Psf` directly currently, then forcing the trampoline would add some unneeded overhead.  Since `Psf` isn't constructible though, I guess this isn't an issue.  One thing I'm not sure about is whether forcing the trampoline on `Psf` would add any overhead to c++-derived subclasses, e.g., `SingleGaussianPsf`.

Reply:
> Interesting. My naive attempt at the factory failed since returning a Psf instance isn't allowed (b/c it's abstract).

Right, I hadn't thought through what the code would actually look like. I guess you could still do it with templates, but that might be a cure worse than the disease.

Hmm... a thought: if `py::init<>()` can access the constructor at all, that means it's probably doing so through the trampoline class (where your using-declaration creates a _public_ constructor with the same signature). So maybe just listing the arguments the naive way will work?

Reply:
Good idea, but no dice.  I tried: 

`cls.def(py::init<bool, std::size_t>(), "init", "isFixed", "capacity");  // Constructor for pure-Python subclasses`

and got a compile error:

```
/software/lsstsw/stack_20200515/python/miniconda3-4.7.12/envs/lsst-scipipe/include/python3.7m/pybind11/detail/init.h:63:64: error: 'lsst::afw::detection::Psf::Psf(bool, std::size_t)' is protected within this context
 inline Class *construct_or_initialize(Args &&...args) { return new Class{std::forward<Args>(args)...}; }
                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from python/lsst/afw/detection/_psf.cc:35:0:
include/lsst/afw/detection/Psf.h:291:14: note: declared protected here
     explicit Psf(bool isFixed = false, std::size_t capacity = 100);
```

Reply:
Actually, I'm confused now as to how _any_ variant of the `using Base::Base` inheriting constructor magic works to change the access modifier of a constructor.  Here's a much simpler similar compile fail without any pybind11: https://godbolt.org/z/Fz2ipy

Reply:
Oops, you're right, there's an exception for constructors: https://en.cppreference.com/w/cpp/language/using_declaration#Inheriting_constructors

Reply:
templating the constructor inheritance might work.... https://godbolt.org/z/7-ucUy

Reply:
Adding

```
    template<typename... Args>
    StorableHelper<Base>(Args... args) : Base(args...) {}
```
to `StorableHelper`

and 

```
    template<typename... Args>
    PsfTrampoline<Base>(Args... args) : lsst::afw::typehandling::StorableHelper<Base>(args...) {}
```

to `PsfTrampoline` seems to work, and the constructor arguments are respected.

Reply:
Awkward, but neither pybind11 users nor python users should notice. I like it.